src/mars.d

```diff
@@ -1575,6 +1577,28 @@ Language changes listed by -transition=id:
             gendocfile(m);
         }
     }
+    if (global.params.vcg_ast)
+    {
+        import ddmd.hdrgen;
+        foreach (mod; modules)
+        {
+            auto buf = new OutBuffer;
+            buf.doindent = 1;
+            scope HdrGenState hgs;
+            hgs.fullDump = 1;
+            scope PrettyPrintVisitor ppv = new PrettyPrintVisitor(buf, &hgs);
+            mod.accept(ppv);
+
+            auto modFilename = mod.srcfile.toChars();
+            auto modFilenameLength = strlen(modFilename);
+            auto cgFilename = cast(char*)allocmemory(modFilenameLength + 4);
+            strcpy(cgFilename, modFilename);
```

Comment:
One reason I dislike ad-hoc C string manipulation is the use of `strcpy` here. It should be `memcpy(cgFilename, modFilename, modFilenameLength)` because the length is already computed, and `strcpy` will compute it again, and `strcpy` will also add a 0 which will get promptly overwritten.